controllers/dexclient_controller.go

```diff
@@ -47,13 +52,143 @@ type DexClientReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *DexClientReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	// _ = log.FromContext(ctx)
 
-	// your logic here
+	log := ctrllog.FromContext(ctx)
+	dexv1Client := &authv1alpha1.DexClient{}
+	if err := r.Get(ctx, req.NamespacedName, dexv1Client); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
+	log.Info("found dexclient", "DexClient.name", dexv1Client.Name, "DexClient.namespace", dexv1Client.Namespace)
+
+	switch {
+
+	// Expect the secret to exist in the current namespace.
+	// The secret has a static name 'mtls-secret-mtls'
+	// If it does not exist, requeue every 2 second
+	// case isNotDefinedMTLSSecretClient(dexv1Client, r, ctx):
+	case isNotDefinedMTLSStruct():
+		log.Info("MTLS values not defined, requeue")
+		return ctrl.Result{RequeueAfter: time.Second}, nil
+
+	// Get the mtls certs from the struct
+	case isDefinedMTLSStruct(dexv1Client, r, ctx):
+		log.Info("MTLS values defined, processing")
+		if r.DexApiClient == nil {
+			dexApiOptions := &dexapi.Options{
+				HostAndPort:    fmt.Sprintf("%s%s", GRPC_SERVICE_NAME, ":5557"), // restruct the grpc service name to `dex`
+				ClientCA:       ctls.caPEM.String(),
+				ClientCrt:      ctls.clientPEM.String(),
+				ClientKey:      ctls.clientPrivKeyPEM.String(),
+				ClientCABytes:  ctls.caPEM.Bytes(),
+				ClientCrtBytes: ctls.clientPEM.Bytes(),
+				ClientKeyBytes: ctls.clientPrivKeyPEM.Bytes(),
+			}
+			dexApiClient, err := dexapi.NewClientPEM(dexApiOptions)
+			if err != nil {
+				log.Error(err, "Api Client create failed", "client", dexv1Client.Name)
+				dexv1Client.Status.State = authv1alpha1.PhaseFailed
+				dexv1Client.Status.Message = err.Error()
+				return ctrl.Result{Requeue: true}, err
+			}
+			// log.Info("created DexApiClient", "caPEM", ctls.caPEM.Bytes(), "clientPEM", ctls.clientPEM.Bytes(), "clientKeyPEM", ctls.clientPrivKeyPEM.Bytes())
+			// log.Info("created DexApiClient", "ca", ctls.caPEM.String(), "clientCrt", ctls.clientPEM.String(), "clientKey", ctls.clientPrivKeyPEM.String())
+			r.DexApiClient = dexApiClient
+		} else {
+			log.Info("found DexApiClient is defined")
+			info, err := r.DexApiClient.GetServerInfo(ctx)
+			if err != nil {
+				log.Info("Failed to get grpc server status:", "error", err)
+				return ctrl.Result{Requeue: true}, err
+			}
+			log.Info("DexApiClient is defined", "grpc server info:", info)
+		}
+		// Drop out
+		// return ctrl.Result{}, nil
```

Comment:
Why is it commented out? I'm lost here..  may you explain the logic?

Reply:
Ok, will refoactor this logic and maybe move into state machine.